_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+higher priority interrupt and, conveniently, upon completion of
+the late arriving exception handler, the processor can then tail-chain into the initial exception
+that was going to be serviced.
+
+### Lazy State Preservation
+
+ARMv7 & ARMv8 devices can implement an optional **Floating Point Unit** (`FPU`) for native floating
+point support. This comes with the addition of 33 4 byte registers (`s0`-`s31` & `fpscr`). 17 of
+these are "caller" saved and need to be dealt with by the ARM exception entry handler. Since FPU
+registers are not often used in ISRs, there is an optimization that can be enabled which defers the
+actual saving of the FPU registers on the stack until a floating point instruction is used in the
+exception. By deferring the actual push of the registers, interrupt latency can usually be reduced by
+the push and pop of these 17 registers!
+
+> A full discussion of FPU stacking optimizations is outside the scope of this article but better managing how the
+> registers are stacked can also be a very helpful tool for reducing stack overflows and memory usage in embedded
+> environments ... Preserving FPU state across context switches requires an additional 132 bytes (33
+> registers) of data to be tracked for each thread! For further reading, ARM wrote a great
+> application note highlighting the lazy preservation FPU features[^8]
+
+#### Execution Priority & Priority Boosting
+
+If no exception is active, the current "execution priority" of the system can be thought of as being the
+"highest configurable priority level" + 1 -- essentially meaning if any exception is pended, the
+currently running code will be interrupted and the ISR will run.
+
+There are a few ways in software that the "execution priority" can be manipulated to be above the
+default priority of thread mode or the exception that is active. This is known as "priority
+boosting". This can be useful to do in software when running code that _cannot_ be interrupted such
+as the logic dealing with context switching in a RTOS.
+
+**Priority boosting** is usually controlled via three register fields:
+
+- **PRIMASK** - Typically configured in code using the CMSIS `__disable_irq()` and `__enable_irq()`
+  routines or the `cpsid i` and `cpsie i` assembly instructions directly. Setting the PRIMASK to 1 disables all interrupts
+  of **configurable priority**. This means, only `NMI`, `Hardfault`, & `Reset` exceptions can still occur.
+- **FAULTMASK** - Typically configured in code using the CMSIS `__disable_fault_irq()` and
+  `__enable_fault_irq()` routines or the `cpsid f` and `cpsie f` assembly instructions directly. Setting the FAULTMASK
+  disables all interrupts except the `NMI` interrupt. This register is not available for ARMv6-M devices.
+- **BASEPRI**- Typically configured using the CMSIS `__set_BASEPRI()` routine. The register has no effect when
+  set to 0 and can be set anywhere from the highest priority level, N, to 1. It can be used to
```

Comment:
I would swap the order of the last two sentence. First "It can be used to prevent ...", then "The register has no effect when set to 0".